inst/include/Rcpp/vector/Subsetter.h

```diff
@@ -132,34 +134,40 @@ class SubsetProxy {
 private:
 
     #ifndef RCPP_NO_BOUNDS_CHECK
-    void check_indices(int* x, int n, int size) {
-        for (int i=0; i < n; ++i) {
+    template <typename IDX>
+    void check_indices(IDX* x, R_xlen_t n, R_xlen_t size) {
+        for (IDX i=0; i < n; ++i) {
             if (x[i] < 0 or x[i] >= size) {
+                if(std::numeric_limits<IDX>::is_integer && size > std::numeric_limits<IDX>::max()) {
+                    stop("use NumericVector to index an object of this size");
```

Comment:
What about `stop("use NumericVector to index an object of length %td", size);`

The error message can be a little nicer.